robot_calibration/include/robot_calibration/ceres/camera_to_camera_error.h

```diff
@@ -0,0 +1,186 @@
+      offsets_ = offsets;
+      data_ = data;
+    }
+
+    virtual ~CameraToCameraError() {}
+
+    /**
+     *  \brief Operator called by CERES optimizer.
+     *  \param free_params The offsets to be applied to joints/transforms.
+     *  \param residuals The residuals computed, to be returned to the optimizer.
+     */
+    bool operator()(double const * const * free_params,
+        double* residuals) const
+    {
+      // Update calibration offsets based on free params
+      offsets_->update(free_params[0]);
+
+      // Project the first camera observations
+      std::vector <geometry_msgs::PointStamped> camera1_pts =
+          camera1_model_->project(data_, *offsets_);
+
+      // Project the second camera estimation
+      std::vector <geometry_msgs::PointStamped> camera2_pts =
+          camera2_model_->project(data_, *offsets_);
+
+      // Calculate plane normals using SVD
+
+      std::vector <cv::Point3f> camera1_points;
+      std::vector <cv::Point3f> camera2_points;
+
+      for (size_t i = 0; i < camera1_pts.size(); ++i)
+      {
+        camera1_points.push_back(cv::Point3f(camera1_pts[i].point.x, camera1_pts[i].point.y, camera1_pts[i].point.z));
+      }
+      for (size_t i = 0; i < camera2_pts.size(); ++i)
+      {
+        camera2_points.push_back(cv::Point3f(camera2_pts[i].point.x, camera2_pts[i].point.y, camera2_pts[i].point.z));
+      }
+
+      // Find the planes
+      cv::Mat plane_1 = getPlane(camera1_points);
+      cv::Mat plane_2 = getPlane(camera2_points);
+
+      // Compute the residuals by minimizing the normals of the calculated planes
+      residuals[0] = std::fabs(plane_1.at<float>(0, 0)) - std::fabs(plane_2.at<float>(0, 0));
+      residuals[1] = std::fabs(plane_1.at<float>(0, 1)) - std::fabs(plane_2.at<float>(0, 1));
+      residuals[2] = std::fabs(plane_1.at<float>(0, 2)) - std::fabs(plane_2.at<float>(0, 2));
+
+      return true;  // always return true
+    }
+
+    /**
+     *  \brief Helper factory function to create a new error block. Parameters
+     *         are described in the class constructor, which this function calls.
+     *  \tparam num_points The number of points in the observation, this forms the
+     *          size of the residuals.
+     *  \tparam num_free_params The number of free parameters being used for
+     *          joint and link calibration.
+     */
+    static ceres::CostFunction* Create(Camera3dModel* camera1_model,
+        Camera3dModel* camera2_model,
+        CalibrationOffsetParser* offsets,
+        robot_calibration_msgs::CalibrationData& data)
+    {
+      int index = -1;
+      for (size_t k = 0; k < data.observations.size() ; k++)
+      {
+        if ( data.observations[k].sensor_name == camera2_model->name())
+        {
+          index = k;
+          break;
+        }
+      }
+
+      if (index == -1)
+      {
+        std::cerr << "Sensor name doesn't match any of the existing finders" << std::endl;
+        return 0;
+      } 
```

Comment:
L119-133 doesn't seem like it should be in this function -- it should be around L272 in optimizer.cpp right after we extract camera1_model -- if we haven't found camera2_model -- we should abort creating the block. returning 0 here give you a null pointer -- which is probably going to blow shit up.

Reply:
I updated the one in optimizer.cpp but the other error blocks also have the check. Should I remove it from them as well??